Mathlib/RingTheory/Bialgebra.lean

```diff
@@ -0,0 +1,70 @@
+/-
+Copyright (c) 2024 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.Coalgebra
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Bialgebras
+
+In this file we define `Bialgebra`, and provide instances for:
+
+* Commutative semirings: `CommSemiring.toBialgebra`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Bialgebra>
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A bialgebra over a commutative (semi)ring `R` is both an algebra and a coalgebra over `R`, such
+that the counit and comultiplication are algebra morphisms. -/
+class Bialgebra (R : Type u) (A : Type v) [CommSemiring R] [Semiring A] extends
+    Algebra R A, Coalgebra R A where
+  /-- The counit is an algebra morphism -/
+  counit_mul : ∀ a₁ a₂ : A, counit (a₁ * a₂) = counit a₁ * counit a₂
+  counit_one : counit 1 = 1
+  /-- The comultiplication is an algebra morphism -/
+  comul_mul : ∀ a₁ a₂ : A, comul (a₁ * a₂) = comul a₁ * comul a₂
+  comul_one : comul 1 = 1
+
+namespace Bialgebra
+variable {R : Type u} {A : Type v}
+variable [CommSemiring R] [Semiring A] [B : Bialgebra R A]
+
+@[simp]
+theorem counit_mul_apply (a₁ a₂ : A) : B.counit (a₁ * a₂) = B.counit a₁ * B.counit a₂ :=
```

Comment:
I think this lemma can be dropped, along with the one below; `apply` seems unnecessary in the lemma name, and you can use `attribute [simp] Bialgebra.comul_mul` instead to avoid duplicating the statement